cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaQuorumCheck.java

```diff
@@ -0,0 +1,110 @@
+
+import static java.lang.Math.ceil;
+
+/**
+ * Provides a method that determines whether it's safe to restart a KRaft controller, which may be the active controller
+ */
+class KafkaQuorumCheck {
+
+    private static final ReconciliationLogger LOGGER = ReconciliationLogger.create(KafkaQuorumCheck.class.getName());
+    private final Reconciliation reconciliation;
+    private final Admin admin;
+    private final Vertx vertx;
+    private final long controllerQuorumFetchTimeoutMs;
+    private Future<QuorumInfo> quorumInfoFuture;
+
+    KafkaQuorumCheck(Reconciliation reconciliation, Admin ac, Vertx vertx, long controllerQuorumFetchTimeoutMs) {
+        this.reconciliation = reconciliation;
+        this.admin = ac;
+        this.vertx = vertx;
+        this.controllerQuorumFetchTimeoutMs = controllerQuorumFetchTimeoutMs;
+    }
+
+    /**
+     * Returns future that completes with true if the given controller can be rolled based on the quorum state. Quorum is considered
+     * healthy if the majority of controllers have caught up with the quorum leader within the controller.quorum.fetch.timeout.ms.
+     */
+    Future<Boolean> canRollController(int podId) {
+        LOGGER.debugCr(reconciliation, "Determining whether controller {} can be rolled", podId);
+        if (quorumInfoFuture == null) {
+            this.quorumInfoFuture = describeMetadataQuorum();
+        }
+        return quorumInfoFuture.map(info -> {
+            boolean canRoll = isQuorumHealthyWithoutPod(podId, info);
+            if (!canRoll) {
+                LOGGER.debugCr(reconciliation, "Restart pod {} would affect the quorum", podId);
+            }
+            return canRoll;
+        }).recover(error -> {
+            LOGGER.warnCr(reconciliation, "Error determining whether it is safe to restart pod {}", podId, error);
+            return Future.failedFuture(error);
+        });
+    }
+
+    /**
+     * Returns id of the quorum leader.
+     **/
+    Future<Integer> quorumLeaderId() {
+        LOGGER.debugCr(reconciliation, "Determining the quorum leader id");
+        if (quorumInfoFuture == null) {
+            this.quorumInfoFuture = describeMetadataQuorum();
+        }
+        return quorumInfoFuture.map(QuorumInfo::leaderId).recover(error -> {
+            LOGGER.warnCr(reconciliation, "Error determining the quorum leader id", error);
+            return Future.failedFuture(error);
+        });
+    }
+
+    private boolean isQuorumHealthyWithoutPod(int podId, QuorumInfo info) {
+        int leaderId = info.leaderId();
+        Map<Integer, Long> replicaStates = info.voters().stream().collect(Collectors.toMap(
+                QuorumInfo.ReplicaState::replicaId,
+                replicaState -> replicaState.lastCaughtUpTimestamp().isPresent() ? replicaState.lastCaughtUpTimestamp().getAsLong() : -1));
+        int totalNumOfControllers = replicaStates.size();
+        AtomicInteger numOfCaughtUpControllers = new AtomicInteger();
```

Comment:
Does this really need to be AtomicInteger? You seem to just loop through it, so normal integer should be sufficient?

Reply:
The reason AtomicInteger is used is because it's being incremented inside lambda function. Alternative to this would be something like:
```
long numOfCaughtUpControllers = replicaStates.entrySet().stream().filter((entry) -> {
            int replicaId = entry.getKey();
            Long lastCaughtUpTimestamp = entry.getValue();
            if (lastCaughtUpTimestamp < 0) {
                LOGGER.warnCr(reconciliation, "No valid lastCaughtUpTimestamp is found for the replica {} ", replicaId);
                return false;
           ...
}).count();
```
Do you prefer this over the AtomicInteger?

Reply:
Ahh, right ... its the _should be final in lambda_ thing. MAybe comment might be useful for people to remember it.